--- a/folia-api/build.gradle.kts
+++ b/folia-api/build.gradle.kts
@@ -51,6 +_,7 @@
     api("org.apache.logging.log4j:log4j-api:$log4jVersion")
     api("org.slf4j:slf4j-api:$slf4jVersion")
     api("com.mojang:brigadier:1.3.10")
+    api("io.sentry:sentry:8.0.0-rc.2") // Pufferfish
 
     // Deprecate bungeecord-chat in favor of adventure
     api("net.md-5:bungeecord-chat:$bungeeCordChatVersion") {
@@ -101,17 +_,21 @@
         java {
             srcDir(generatedDir)
             srcDir(file("../paper-api/src/main/java"))
+            srcDir(file("../folia-api/src/main/java"))
         }
         resources {
             srcDir(file("../paper-api/src/main/resources"))
+            srcDir(file("../folia-api/src/main/resources"))
         }
     }
     test {
         java {
             srcDir(file("../paper-api/src/test/java"))
+            srcDir(file("../folia-api/src/test/java"))
         }
         resources {
             srcDir(file("../paper-api/src/test/resources"))
+            srcDir(file("../folia-api/src/test/resources"))
         }
     }
 }
@@ -257,3 +_,10 @@
         dependsOn(scanJarForOldGeneratedCode)
     }
 }
+
+// Pufferfish Start
+tasks.withType<JavaCompile> {
+    val compilerArgs = options.compilerArgs
+    compilerArgs.add("--add-modules=jdk.incubator.vector")
+}
+// Pufferfish End
