From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: MC_XiaoHei <xiaohei.xor7@outlook.com>
Date: Sun, 21 Jul 2024 00:06:21 +0800
Subject: [PATCH] Command API support


diff --git a/src/main/java/org/bukkit/plugin/java/JavaPlugin.java b/src/main/java/org/bukkit/plugin/java/JavaPlugin.java
index d5a3c3dce76c4ed0f1184ab5ba21db9c5f1c01ec..6f968a774bdbbb672fe5df8ba6ed8fb499e4dfde 100644
--- a/src/main/java/org/bukkit/plugin/java/JavaPlugin.java
+++ b/src/main/java/org/bukkit/plugin/java/JavaPlugin.java
@@ -15,6 +15,8 @@ import java.util.List;
 import java.util.Locale;
 import java.util.logging.Level;
 import java.util.logging.Logger;
+
+import org.bukkit.Bukkit;
 import org.bukkit.Server;
 import org.bukkit.command.Command;
 import org.bukkit.command.CommandSender;
@@ -64,6 +66,17 @@ public abstract class JavaPlugin extends PluginBase {
         // Paper end
     }
 
+    // Lumina start - CommandAPI support
+    // SHOULD NOT BE USED IN PLUGINS !!!
+    public JavaPlugin(@Nullable Object onlyForCommandApi) {
+        server = Bukkit.getServer();
+        isEnabled = true;
+        description = new PluginDescriptionFile("CommandAPI", "", "");
+        logger = Logger.getLogger("CommandAPI");
+        pluginMeta = description;
+    }
+    // Lumina end - CommandAPI support
+
     @Deprecated(forRemoval = true) // Paper
     protected JavaPlugin(@NotNull final JavaPluginLoader loader, @NotNull final PluginDescriptionFile description, @NotNull final File dataFolder, @NotNull final File file) {
         final ClassLoader classLoader = this.getClass().getClassLoader();
