From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: MC_XiaoHei <xiaohei.xor7@outlook.com>
Date: Sat, 3 Aug 2024 12:24:14 +0800
Subject: [PATCH] FakePlayer API


diff --git a/src/main/java/org/leavesmc/lumina/entity/FakePlayer.java b/src/main/java/org/leavesmc/lumina/entity/FakePlayer.java
new file mode 100644
index 0000000000000000000000000000000000000000..d79a3be1e73317641121e2a2edc7abbfb0929d82
--- /dev/null
+++ b/src/main/java/org/leavesmc/lumina/entity/FakePlayer.java
@@ -0,0 +1,26 @@
+package org.leavesmc.lumina.entity;
+
+public interface FakePlayer {
+
+    void startAction(ActionType actionType, int limit, int interval, boolean continuous);
+
+    default void startActionOnce(ActionType actionType) {
+        startAction(actionType, 1, 1, false);
+    }
+
+    default void startActionTimes(ActionType actionType, int times) {
+        startAction(actionType, times, 1, false);
+    }
+
+    default void startActionContinuous(ActionType actionType) {
+        startAction(actionType, -1, 1, true);
+    }
+
+    default void startActionInterval(ActionType actionType, int interval) {
+        startAction(actionType, -1, interval, false);
+    }
+
+    enum ActionType {
+        USE, ATTACK, JUMP, DROP_ITEM, DROP_STACK, SWAP_HANDS
+    }
+}
diff --git a/src/main/java/org/leavesmc/lumina/utils/PlayerType.java b/src/main/java/org/leavesmc/lumina/utils/PlayerType.java
new file mode 100644
index 0000000000000000000000000000000000000000..c8f8f3ab26bbfe9c52f02659d642281e59ed7b48
--- /dev/null
+++ b/src/main/java/org/leavesmc/lumina/utils/PlayerType.java
@@ -0,0 +1,7 @@
+package org.leavesmc.lumina.utils;
+
+public enum PlayerType {
+    REAL,
+    FAKE,
+    GHOST
+}
\ No newline at end of file
