From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: MC_XiaoHei <xiaohei.xor7@outlook.com>
Date: Wed, 24 Jul 2024 04:01:45 +0000
Subject: [PATCH] Change default value of some configs for technical


diff --git a/src/main/java/io/papermc/paper/configuration/GlobalConfiguration.java b/src/main/java/io/papermc/paper/configuration/GlobalConfiguration.java
index 4ed27c10b432ceebf4447ab8007bc3a1be09a06e..8193a969540adfaac8098a1408c67398d10e2397 100644
--- a/src/main/java/io/papermc/paper/configuration/GlobalConfiguration.java
+++ b/src/main/java/io/papermc/paper/configuration/GlobalConfiguration.java
@@ -165,14 +165,14 @@ public class GlobalConfiguration extends ConfigurationPart {
 
     public class UnsupportedSettings extends ConfigurationPart {
         @Comment("This setting allows for exploits related to end portals, for example sand duping")
-        public boolean allowUnsafeEndPortalTeleportation = false;
+        public boolean allowUnsafeEndPortalTeleportation = true; // Lumina - change config default value for technical
         @Comment("This setting controls if players should be able to break bedrock, end portals and other intended to be permanent blocks.")
-        public boolean allowPermanentBlockBreakExploits = false;
+        public boolean allowPermanentBlockBreakExploits = true; // Lumina - change config default value for technical
         @Comment("This setting controls if player should be able to use TNT duplication, but this also allows duplicating carpet, rails and potentially other items")
-        public boolean allowPistonDuplication = false;
+        public boolean allowPistonDuplication = true; // Lumina - change config default value for technical
         public boolean performUsernameValidation = true;
         @Comment("This setting controls if players should be able to create headless pistons.")
-        public boolean allowHeadlessPistons = false;
+        public boolean allowHeadlessPistons = true; // Lumina - change config default value for technical
         @Comment("This setting controls if the vanilla damage tick should be skipped if damage was blocked via a shield.")
         public boolean skipVanillaDamageTickWhenShieldBlocked = false;
         @Comment("This setting controls what compression format is used for region files.")
@@ -190,7 +190,7 @@ public class GlobalConfiguration extends ConfigurationPart {
     public class Commands extends ConfigurationPart {
         public boolean suggestPlayerNamesWhenNullTabCompletions = true;
         public boolean fixTargetSelectorTagCompletion = true;
-        public boolean timeCommandAffectsAllWorlds = false;
+        public boolean timeCommandAffectsAllWorlds = true; // Lumina - change config default value for technical
     }
 
     public Logging logging;
@@ -256,15 +256,15 @@ public class GlobalConfiguration extends ConfigurationPart {
             public IntOr.Disabled pageMax = new IntOr.Disabled(OptionalInt.of(2560)); // TODO this appears to be a duplicate setting with one above
             public double totalMultiplier = 0.98D; // TODO this should probably be merged into the above inner class
         }
-        public boolean resolveSelectorsInBooks = false;
+        public boolean resolveSelectorsInBooks = true; // Lumina - change config default value for technical
     }
 
     public PacketLimiter packetLimiter;
 
     public class PacketLimiter extends ConfigurationPart {
         public Component kickMessage = Component.translatable("disconnect.exceeded_packet_rate", NamedTextColor.RED);
-        public PacketLimit allPackets = new PacketLimit(7.0, 500.0, PacketLimit.ViolateAction.KICK);
-        public Map<Class<? extends Packet<?>>, PacketLimit> overrides = Map.of(ServerboundPlaceRecipePacket.class, new PacketLimit(4.0, 5.0, PacketLimit.ViolateAction.DROP));
+        public PacketLimit allPackets = new PacketLimit(1.0, 114514.0, PacketLimit.ViolateAction.KICK); // Lumina - change config default value for technical
+        public Map<Class<? extends Packet<?>>, PacketLimit> overrides = Map.of(); // Lumina - change config default value for technical
 
         @ConfigSerializable
         public record PacketLimit(@Required double interval, @Required double maxPacketRate, ViolateAction action) {
@@ -333,7 +333,7 @@ public class GlobalConfiguration extends ConfigurationPart {
             }
         }
         public int maxJoinsPerTick = 5;
-        public boolean fixEntityPositionDesync = true;
+        public boolean fixEntityPositionDesync = false; // Lumina - change config default value for technical
         public boolean loadPermissionsYmlBeforePlugins = true;
         @Constraints.Min(4)
         public int regionFileCacheSize = 256;
diff --git a/src/main/java/io/papermc/paper/configuration/WorldConfiguration.java b/src/main/java/io/papermc/paper/configuration/WorldConfiguration.java
index ad3d9251e88ce9a35b0f46e9f0f54e62559bd9f9..7bc4d2f2015346aa3bb36b80ac5b1ba990411aaf 100644
--- a/src/main/java/io/papermc/paper/configuration/WorldConfiguration.java
+++ b/src/main/java/io/papermc/paper/configuration/WorldConfiguration.java
@@ -179,10 +179,10 @@ public class WorldConfiguration extends ConfigurationPart {
         public class Spawning extends ConfigurationPart {
             public ArrowDespawnRate nonPlayerArrowDespawnRate = ArrowDespawnRate.def(WorldConfiguration.this.spigotConfig);
             public ArrowDespawnRate creativeArrowDespawnRate = ArrowDespawnRate.def(WorldConfiguration.this.spigotConfig);
-            public boolean filterBadTileEntityNbtFromFallingBlocks = true;
-            public List<NbtPathArgument.NbtPath> filteredEntityTagNbtPaths = NbtPathSerializer.fromString(List.of("Pos", "Motion", "SleepingX", "SleepingY", "SleepingZ"));
+            public boolean filterBadTileEntityNbtFromFallingBlocks = false; // Lumina - change config default value for technical
+            public List<NbtPathArgument.NbtPath> filteredEntityTagNbtPaths = NbtPathSerializer.fromString(List.of()); // Lumina - change config default value for technical
             public boolean disableMobSpawnerSpawnEggTransformation = false;
-            public boolean perPlayerMobSpawns = true;
+            public boolean perPlayerMobSpawns = false; // Lumina - change config default value for technical
             public boolean scanForLegacyEnderDragon = true;
             @MergeMap
             public Reference2IntMap<MobCategory> spawnLimits = Util.make(new Reference2IntOpenHashMap<>(NaturalSpawner.SPAWNING_CATEGORIES.length), map -> Arrays.stream(NaturalSpawner.SPAWNING_CATEGORIES).forEach(mobCategory -> map.put(mobCategory, -1)));
@@ -257,13 +257,13 @@ public class WorldConfiguration extends ConfigurationPart {
             @BelowZeroToEmpty
             public DoubleOr.Default skeletonHorseThunderSpawnChance = DoubleOr.Default.USE_DEFAULT;
             public boolean ironGolemsCanSpawnInAir = false;
-            public boolean countAllMobsForSpawning = false;
+            public boolean countAllMobsForSpawning = true; // Lumina - change config default value for technical
             @BelowZeroToEmpty
             public IntOr.Default monsterSpawnMaxLightLevel = IntOr.Default.USE_DEFAULT;
             public DuplicateUUID duplicateUuid;
 
             public class DuplicateUUID extends ConfigurationPart {
-                public DuplicateUUIDMode mode = DuplicateUUIDMode.SAFE_REGEN;
+                public DuplicateUUIDMode mode = DuplicateUUIDMode.WARN; // Lumina - change config default value for technical
                 public int safeRegenDeleteRange = 32;
 
                 public enum DuplicateUUIDMode {
@@ -301,8 +301,8 @@ public class WorldConfiguration extends ConfigurationPart {
 
             public boolean disableCreeperLingeringEffect = false;
             public boolean enderDragonsDeathAlwaysPlacesDragonEgg = false;
-            public boolean phantomsDoNotSpawnOnCreativePlayers = true;
-            public boolean phantomsOnlyAttackInsomniacs = true;
+            public boolean phantomsDoNotSpawnOnCreativePlayers = false; // Lumina - change config default value for technical
+            public boolean phantomsOnlyAttackInsomniacs = false; // Lumina - change config default value for technical
             public int playerInsomniaStartTicks = 72000;
             public int phantomsSpawnAttemptMinSeconds = 60;
             public int phantomsSpawnAttemptMaxSeconds = 119;
@@ -406,7 +406,7 @@ public class WorldConfiguration extends ConfigurationPart {
 
     public class Scoreboards extends ConfigurationPart {
         public boolean allowNonPlayerEntitiesOnScoreboards = true;
-        public boolean useVanillaWorldScoreboardNameColoring = false;
+        public boolean useVanillaWorldScoreboardNameColoring = true; // Lumina - change config default value for technical
     }
 
     public Environment environment;
@@ -461,7 +461,7 @@ public class WorldConfiguration extends ConfigurationPart {
     public Maps maps;
 
     public class Maps extends ConfigurationPart {
-        public int itemFrameCursorLimit = 128;
+        public int itemFrameCursorLimit = 10086; // Lumina - change config default value for technical
         public int itemFrameCursorUpdateInterval = 10;
     }
 
@@ -479,14 +479,14 @@ public class WorldConfiguration extends ConfigurationPart {
     public UnsupportedSettings unsupportedSettings;
 
     public class UnsupportedSettings extends ConfigurationPart {
-        public boolean fixInvulnerableEndCrystalExploit = true;
+        public boolean fixInvulnerableEndCrystalExploit = false; // Lumina - change config default value for technical
         public boolean disableWorldTickingWhenEmpty = false;
     }
 
     public Hopper hopper;
 
     public class Hopper extends ConfigurationPart {
-        public boolean cooldownWhenFull = true;
+        public boolean cooldownWhenFull = false; // Lumina - change config default value for technical
         public boolean disableMoveEvent = false;
         public boolean ignoreOccludingBlocks = false;
     }
@@ -498,8 +498,8 @@ public class WorldConfiguration extends ConfigurationPart {
         public boolean allowVehicleCollisions = true;
         public boolean fixClimbingBypassingCrammingRule = false;
         @RequiresSpigotInitialization(MaxEntityCollisionsInitializer.class)
-        public int maxEntityCollisions = 8;
-        public boolean allowPlayerCrammingDamage = false;
+        public int maxEntityCollisions = 114514; // Lumina - change config default value for technical
+        public boolean allowPlayerCrammingDamage = true; // Lumina - change config default value for technical
     }
 
     public Chunks chunks;
@@ -514,10 +514,10 @@ public class WorldConfiguration extends ConfigurationPart {
         // Folia end - region threading - force prevent moving into unloaded chunks
 
         public AutosavePeriod autoSaveInterval = AutosavePeriod.def();
-        public int maxAutoSaveChunksPerTick = 24;
+        public int maxAutoSaveChunksPerTick = 200; // Lumina - change config default value for technical
         public int fixedChunkInhabitedTime = -1;
         public boolean preventMovingIntoUnloadedChunks = false;
-        public Duration delayChunkUnloadsBy = Duration.of("10s");
+        public Duration delayChunkUnloadsBy = Duration.of("0s"); // Lumina - change config default value for technical
         public Reference2IntMap<EntityType<?>> entityPerChunkSaveLimit = Util.make(new Reference2IntOpenHashMap<>(BuiltInRegistries.ENTITY_TYPE.size()), map -> {
             map.defaultReturnValue(-1);
             map.put(EntityType.EXPERIENCE_ORB, -1);
diff --git a/src/main/java/org/spigotmc/SpigotWorldConfig.java b/src/main/java/org/spigotmc/SpigotWorldConfig.java
index 9f38a0763597d9d70cb8d1b636c7d4b14d32c535..80dfcdcedb7beda80db9deec0c40201ce36fc9c8 100644
--- a/src/main/java/org/spigotmc/SpigotWorldConfig.java
+++ b/src/main/java/org/spigotmc/SpigotWorldConfig.java
@@ -436,14 +436,14 @@ public class SpigotWorldConfig
     }
 
     //public int currentPrimedTnt = 0; // Folia - region threading - moved to regionised world data
-    public int maxTntTicksPerTick;
+    public int maxTntTicksPerTick = 0; // Lumina - change config default value for technical
     private void maxTntPerTick() {
         if ( SpigotConfig.version < 7 )
         {
             this.set( "max-tnt-per-tick", 100 );
         }
-        this.maxTntTicksPerTick = this.getInt( "max-tnt-per-tick", 100 );
-        this.log( "Max TNT Explosions: " + this.maxTntTicksPerTick );
+        this.maxTntTicksPerTick = this.getInt( "max-tnt-per-tick", 0 ); // Lumina - change config default value for technical
+        this.log( "Max TNT Explosions: " + (this.maxTntTicksPerTick == 0 ? "infinity" : this.maxTntTicksPerTick )); // Lumina - better logging
     }
 
     public int hangingTickFrequency;
