From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: HaHaWTH <fsjk947@gmail.com>
Date: Fri, 15 Mar 2024 03:32:46 +0800
Subject: [PATCH] Luminol Fix-MC-2025


diff --git a/src/main/java/net/minecraft/world/entity/Entity.java b/src/main/java/net/minecraft/world/entity/Entity.java
index f8daab8b7e56421aeabb36e571a9f63d75e4bb73..657a0e0dc8350704d89ab9722a207a65f5f58455 100644
--- a/src/main/java/net/minecraft/world/entity/Entity.java
+++ b/src/main/java/net/minecraft/world/entity/Entity.java
@@ -2602,6 +2602,17 @@ public abstract class Entity implements SyncedDataHolder, Nameable, EntityAccess
                 nbttagcompound.putBoolean("Paper.FreezeLock", true);
             }
             // Paper end
+
+            // Luminol start - Fix MC-2025
+            if(!org.leavesmc.lumina.config.LuminaConfig.configModule.fix.mc2025) return nbttagcompound;
+            AABB boundingBox = getBoundingBox();
+            ListTag boundingBoxList = new ListTag();
+            for (double coord : new double[]{boundingBox.minX, boundingBox.minY, boundingBox.minZ, boundingBox.maxX, boundingBox.maxY, boundingBox.maxZ}) {
+                boundingBoxList.add(DoubleTag.valueOf(coord));
+            }
+            nbttagcompound.put("Luminol.BoundingBox", boundingBoxList);
+            // Luminol end
+
             return nbttagcompound;
         } catch (Throwable throwable) {
             CrashReport crashreport = CrashReport.forThrowable(throwable, "Saving entity NBT");
@@ -2679,6 +2690,13 @@ public abstract class Entity implements SyncedDataHolder, Nameable, EntityAccess
                         this.reapplyPosition();
                     }
 
+                    // Luminol start - Fix MC-2025
+                    if (org.leavesmc.lumina.config.LuminaConfig.configModule.fix.mc2025 && nbt.contains("Luminol.BoundingBox", net.minecraft.nbt.Tag.TAG_LIST)) {
+                        ListTag boundingBoxList = nbt.getList("Luminol.BoundingBox", net.minecraft.nbt.Tag.TAG_DOUBLE);
+                        setBoundingBox(new AABB(boundingBoxList.getDouble(0), boundingBoxList.getDouble(1), boundingBoxList.getDouble(2), boundingBoxList.getDouble(3), boundingBoxList.getDouble(4), boundingBoxList.getDouble(5)));
+                    }
+                    // Luminol end
+
                 } else {
                     throw new IllegalStateException("Entity has invalid rotation");
                 }
