From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: MrHua269 <novau233@163.com>
Date: Wed, 7 Feb 2024 06:30:03 +0000
Subject: [PATCH] Luminol Try fixing folia spector teleportation


diff --git a/src/main/java/net/minecraft/server/level/ServerPlayer.java b/src/main/java/net/minecraft/server/level/ServerPlayer.java
index cbc89b8898f9243776af88c2b04ad3c2e4e23fb6..5b168901b53c523c3fa41867be88f810ea0e6728 100644
--- a/src/main/java/net/minecraft/server/level/ServerPlayer.java
+++ b/src/main/java/net/minecraft/server/level/ServerPlayer.java
@@ -841,6 +841,11 @@ public class ServerPlayer extends Player {
         }
 
         Entity entity = this.getCamera();
+        // Luminol start - Fix folia spector teleportation
+        if (!io.papermc.paper.util.TickThread.isTickThreadFor(entity) && org.leavesmc.lumina.config.modules.fixes.FoliaSpectorTeleportationFixConfig.fixSpectorTeleportFolia){
+            this.setCamera(this);
+        }
+        // Luminol end
 
         if (entity != this) {
             if (entity.isAlive()) {
