From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: MC_XiaoHei <xiaohei.xor7@outlook.com>
Date: Wed, 14 Aug 2024 21:51:45 +0800
Subject: [PATCH] Rewrite Nether Portal logic


diff --git a/src/main/java/net/minecraft/world/entity/ai/village/poi/PoiSection.java b/src/main/java/net/minecraft/world/entity/ai/village/poi/PoiSection.java
index 4ee7d75c56d9f9ff3607276857dde84410ba3f2a..8731cebf72febd84822284914bbf76075e6b4b0b 100644
--- a/src/main/java/net/minecraft/world/entity/ai/village/poi/PoiSection.java
+++ b/src/main/java/net/minecraft/world/entity/ai/village/poi/PoiSection.java
@@ -78,6 +78,7 @@ public class PoiSection {
     private boolean add(PoiRecord poi) {
         BlockPos blockPos = poi.getPos();
         Holder<PoiType> holder = poi.getPoiType();
+        if (holder.is(PoiTypes.NETHER_PORTAL)) System.out.println("Adding nether portal at " + blockPos); // Lumina - rewrite nether portal logic
         short s = SectionPos.sectionRelativePos(blockPos);
         PoiRecord poiRecord = this.records.get(s);
         if (poiRecord != null) {
@@ -95,6 +96,7 @@ public class PoiSection {
 
     public void remove(BlockPos pos) {
         PoiRecord poiRecord = this.records.remove(SectionPos.sectionRelativePos(pos));
+        if (poiRecord.getPoiType().is(PoiTypes.NETHER_PORTAL)) System.out.println("Adding nether portal at " + pos); // Lumina - rewrite nether portal logic
         if (poiRecord == null) {
             LOGGER.error("POI data mismatch: never registered at {}", pos);
         } else {
